### PR TITLE
Always show BlackBox parameters in alphabetical key order

### PR DESCRIPTION
**Pull request by chick**
Always show BlackBox parameters in alphabetical key order
Includes test

### Contributor Checklist

- [x] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- backend code generation

#### API Impact

No API change

#### Backend Code Generation Impact

Verilog is changed
Verilog generation of blackboxes will have the parameter map ordered alphabetically by key

#### Desired Merge Strategy

- Squash: The PR will be squashed and merged (choose this if you have no preference.

#### Release Notes

Verilog is changed
Verilog generation of blackboxes will have the parameter map ordered alphabetically by key

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: src/test/scala/chiselTests/BlackBox.scala [hunk 1]
```diff
@@ -222,4 +230,48 @@ class BlackBoxSpec extends ChiselFlatSpec {
...
   }
+
+  "BlackBoxes" should "sort the verilog output of their param map by param key" in {
+    val rand = scala.util.Random
+    rand.setSeed(0)
+
+    class Top(m: Map[String, Param]) extends Module {
+      val io = IO(new Bundle {})
+      val ibufds = Module(new ParameterizedBlackBox(m))
+      ibufds.io.I := clock
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/BlackBox.scala:241]: I'm a little confused why FIRRTL is letting you only drive one of the inputs but not the other.

File: src/test/scala/chiselTests/BlackBox.scala [hunk 2]
```diff
@@ -222,4 +230,48 @@ class BlackBoxSpec extends ChiselFlatSpec {
...
+      val paramsShowInOrder = verilogString.split("\n").exists { line =>
+        val offset = line.indexOf(".key_0")
+        val result = if (offset >= 0) {
+          paramsSeen = true
+          keysInOrder(line.drop(offset))
+        } else {
+          false
+        }
+        result
+      } || !paramsSeen
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/BlackBox.scala:272]: I find this logic pretty hard to follow. This is a test so it doesn't matter _that_ much but I think it would be more clear and still sufficient to just take a static list of like 6 parameters like parameter named `A` has value `0`, `B` has value `1`, etc., randomize there order, and then just check the exactly expected line in the Verilog, something like:

```scala
verilogString should include ("#(.A(0), .B(1), .C(2)...")
```

**Review by jackkoenig — COMMENTED** [review]
Solution looks great but I find the test a little hard to follow.

File: src/test/scala/chiselTests/BlackBox.scala (hunk 2 above)
Comment (chick) [reply src/test/scala/chiselTests/BlackBox.scala]: Makes sense, I've simplified the test

**Comment by mwachs5** [comment]
looks good! Should we backport this even though it changes the verilog, as it is basically a bug?

**Comment by jackkoenig** [comment]
> looks good! Should we backport this even though it changes the verilog, as it is basically a bug?

This is a tough one, because even though the previous output is not _ordered_ and the order will scramble if you add another parameter, it was at least stable between runs. I'm inclined to not backport this but am open to being convinced.
